transformer_engine/pytorch/attention.py

```diff
@@ -797,6 +810,54 @@ def forward(self, max_seq_len: int, offset: int = 0):
         # emb [seq_length, .., dim]
         return emb.reshape(emb.size(0), 1, 1, emb.size(1))
 
+
+class FusedRoPEFunc(torch.autograd.Function):
+    """
+    Function for FusedRoPE
+
+    This implementation assumes the input tensor to be in `sbhd` or `thd` format and
+    the RoPE tensor to be of shape (s, 1, 1, d). It accepts arbitrary memory layouts to avoid
```

Comment:
How difficult will it be to adapt this to a RoPE tensor `freq` of shape `(1, s, s, d)` and the input tensor `t` in `bshd` format? 
Basically, I wanna make RoPE work for both `sbhd` and `bshd` formats.

Reply:
For `bshd` format, we could transpose the input tensor while keeping `freqs` of shape `(s, 1, 1, d)`. I'll push a version for your review.

Reply:
I think input tensor might be bigger than the `freqs` tensor, so transposing the former might incur some performance cost

Reply:
`transpose` just changes the layout of a tensor but won't do the real transposing in memory. And luckily, this fused rope kernel accepts arbitrary layouts, so ideally it won't introduce any extra cost.

Reply:
Checked via nsys that there's no extra cost.